hydra_base/lib/data.py

```diff
@@ -1092,3 +1095,53 @@ def delete_dataset(dataset_id,**kwargs):
 
 def read_json(json_string):
     pd.read_json(json_string)
+
+def get_hdf_filesize(url, **kwargs):
+    """
+      Returns the size in bytes of the hdf file <url> argument
+      Raises ValueError on bad url
+    """
+    hdf = HdfStorageAdapter()
+    return hdf.size(url)
+
+def get_hdf_dataset_info(url, dataset_name, **kwargs):
+    """
+      Returns information about the <dataset_name> argument:
+      {
+        "name": str: Name of the dataset,
+        "size": int: Number of rows in dataset series,
+        "dtype": str: The numpy.dtype of the dataset
+      }
+      Raises ValueError on bad url or dataset name
+    """
+    hdf = HdfStorageAdapter()
+    return hdf.get_dataset_info_url(url, dataset_name)
+
+def get_hdf_dataframe(url, dataset_name, start, end, **kwargs):
+    """
+      Returns the rows from <start> to <end> of <dataset_name> in
+      the hdf file <url> as the json representation of a Pandas
+      DataFrame. This may then be read directly in a client with
+      pandas.read_json().
+      Raises ValueError on bad url, dataset name or bounds
+    """
+    hdf = HdfStorageAdapter()
```

Comment:
Should there be user ID validation here to check the user can access the data?

Reply:
There should definitely be validation that the user is permitted to access the HDF file *somewhere* in the chain between a client requesting a Hydra dataset and the contents of an HDF file being returned.

At the moment, there is the usual validation when the Hydra dataset which contains a reference to an HDF file is loaded, so if the user can't see the Hydra dataset, they also can't see the HDF url.

There is currently the potential for a user to edit the HDF url in a Hydra dataset which they have access to so that it points to any arbitrary HDF file.  Because they can see the Hydra dataset, they can then see the HDF file.

A minimal improvement here is checking that the HDF url is within an "authorised path", e.g. it goes to "s3://hydra/... rather than "s3://arbitrary-third-party/..." but then we need to pre-register paths and/or ensure that all HDF files are only ever in "authorised" places.  An easy but restrictive way to do this would be to extend the current "filestore" mechanism to require that any HDF file must be in it, ie. no external internet access.

Or particular users can be "HDF enabled" and only they have access to the HDF api.

The other end of the spectrum is per-user per-HDF access control, i.e. keeping a record of which HDF files are accessible to each user *in addition* to which Hydra datasets are accessible to each user.  The user may be able to see the Hydra dataset, but necessarily see the HDF file it references.  This solves the issue of a user rewriting the HDF url in a Hydra dataset which they can access, but is quite a heavyhanded approach.

So in summary, yes there are a few ways access to HDF files can be tightened rather than just having it piggyback on Hydra dataset validation as at present, but there are policy and admin implications for each.